### PR TITLE
Copy over fixtures from ember-suave

### PR DESCRIPTION
**Pull request by rwwagner90**

### CODE REVIEW

**Comment by rwwagner90** [comment]
@brzpegasus this is a WIP, copied over all the fixtures blindly from ember-suave, and we have several failures. I'll work on resolving those failures now.

**Comment by rwwagner90** [comment]
@alexlafroscia @brzpegasus it seems that the ember-suave rule requires no spaces in generators before the `*` and can either have a space or no space after it. Is this correct? Could we decide on space or no space after `*`? ESLint cannot do an optional space after the star.

**Comment by brzpegasus** [comment]
I believe I left the space after `*` unspecified because we want to allow the following:

```js
// good
function* foo() {} // space required after *
let bar = function*() {} // no space after *, for consistency with normal anonymous functions
```

The anonymous function restriction is technically enforced by the rule that forbids spaces before parentheses for anonymous functions:

```js
// good
let foo = function() {}

// bad
let bar = function () {}
```

**Comment by rwwagner90** [comment]
@brzpegasus yeah, I agree with the way rules were set up. It seems ESLint won't allow an optional space after `*` though. It's either require no space or require a space. I posted an issue to change the rule, hopefully.

**Comment by rwwagner90** [comment]
What about `catch()` vs `catch ()`? Seems to be conflicting rules here

**Comment by rwwagner90** [comment]
@brzpegasus correct me if I am wrong, but the fixtures good/bad examples should adhere to all other rules, **except** the one we are testing right?

**Comment by brzpegasus** [comment]
> What about catch() vs catch ()? Seems that space after keyword and space before keyword good/bad files conflict. Not sure how tests passed before

Looks like we never had a rule for it 😛. DockYarders favor `catch ()`. Some of the "good" fixtures might have to be updated.

**Comment by brzpegasus** [comment]
> the fixtures good/bad examples should adhere to all other rules, except the one we are testing right?

Confirm.

**Comment by brzpegasus** [comment]
> DockYarders favor `catch ()`

I take it back. Should be `catch()`, similar to `function() {}`.

**Comment by rwwagner90** [comment]
@brzpegasus one of the big things that seemed to be wrong in all the examples was array spacing. Are we supposed to enforce it or not?

I've got it down to just 4 errors now, one of which is a flaw with ESLint and not sure why they other 3 do not work.

Please take a look at what I have done and let me know if anything seems off.

**Comment by rwwagner90** [comment]
Seems like, for some reason, `newline-after-var` only works for `vars` and not `let` or `const`, but the rule docs say it should work for everything.

**Comment by rwwagner90** [comment]
@brzpegasus @alexlafroscia have either of you had that issue with `newline-after-var` before?

**Comment by brzpegasus** [comment]
In `fixtures/const-usage/good/example.js`, I had to temporarily comment out a `const` statement because there was no new line after it per the `newline-after-var` rule, so it is definitely recognizing `const`.

**Comment by brzpegasus** [comment]
> one of the big things that seemed to be wrong in all the examples was array spacing. Are we supposed to enforce it or not?

If by "wrong", you mean that arrays are formatted as `['a', 'b', 'c']` instead of `[ 'a', 'b', 'c' ]`, then no, that is intentional. We discussed it internally [here](https://github.com/DockYard/styleguides/issues/45).

**Comment by rwwagner90** [comment]
@brzpegasus it doesn't seem to work for the const.js and let.js newline rule ones though :(. As for array spacing, someone enforced array spacing in the config, so I assumed it was supposed to be done. I will change it to enforce **no** space in arrays instead.

**Comment by rwwagner90** [comment]
Okay, I fixed the array spacing to be the correct one. @brzpegasus  you have any idea why `require-line-break-after-variable-assignment/bad/const.js` and `require-line-break-after-variable-assignment/bad/let.js` are not failing like they should? It seems they work, if I change them to `var` instead.

**Comment by brzpegasus** [comment]
> Okay, I fixed the array spacing to be the correct one.

Thanks! I didn't realize we had that many files with the incorrect array spacing.

> you have any idea why require-line-break-after-variable-assignment/bad/const.js and require-line-break-after-variable-assignment/bad/let.js are not failing like they should?

I will check.

**Comment by rwwagner90** [comment]
@brzpegasus you didn't have any incorrect array spacing files. The problem was this plugin initially had the wrong array spacing rule defined, and I assumed all the rules were correct when I started.

**Comment by brzpegasus** [comment]
Ah, gotcha.

For `const.js` and `let.js`, which rule is supposed to disallow `let abc = 1; let foo = 2;`? I don't believe [`newline-after-var`](http://eslint.org/docs/rules/newline-after-var) is intended to address that scenario. It only checks whether you have an empty line after the last variable declaration.

I think when you change `const` and `let` to `var` in those files, the tests report that those files are failing due to the `no-var` rule, not because it somehow works with `var`. Not sure that there's an equivalent in ESLint, but we might be able to achieve the same results with [`max-statements-per-line`](http://eslint.org/docs/rules/max-statements-per-line).

**Comment by rwwagner90** [comment]
Ah, good catch. I will try that now

**Comment by rwwagner90** [comment]
@brzpegasus `max-statements-per-line` did the trick! 😄 . Now we just have 3 issues. I believe all of them are known limitations with ESLint, so not sure we can do any further, without writing some more custom rules or proposing changes to ESLint. 

I proposed the change to ESLint for the generator issue here: https://github.com/eslint/eslint/issues/6777

**Comment by rwwagner90** [comment]
@brzpegasus since this PR gets us most of the way there, what would you think about merging it and opening separate issues for the remaining 3, since they will take some work?

File: tests/fixtures/disallow-multiple-var-decl/good/for.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable ember-suave/prefer-destructuring */
```
Comment (brzpegasus) [inline tests/fixtures/disallow-multiple-var-decl/good/for.js]: Hmm, I don't believe the old custom rule for object destructuring complained about this, though I guess it makes sense that it should per the rule.

File: tests/fixtures/require-template-strings-for-concatenation/good/example.js [hunk 2]
```diff
@@ -3,5 +3,5 @@ let bar = `${foo} stuff`;
 
 foo(`${bar}-${foo}`);
 
-let reallyLong = 'asdfadfdsfasdfadfadfasdf' +
-      'adsfasdf';
+let reallyLong = 'asdfadfdsfasdfadfadfasdf'
+    +  'adsfasdf';
```
Comment (brzpegasus) [inline tests/fixtures/require-template-strings-for-concatenation/good/example.js]: This is new. Which rule is enforcing this?

File: tests/fixtures/require-space-after-keywords/good/example.js [hunk 3]
```diff
@@ -21,12 +21,12 @@ switch (x) {
 
 try {
   x();
-} catch(error) {
+} catch (error) {
```
Comment (brzpegasus) [inline tests/fixtures/require-space-after-keywords/good/example.js]: Sorry for the confusion. I had retracted my comment about `catch` [here](https://github.com/DockYard/eslint-plugin-ember-suave/pull/11#issuecomment-235670286). There's an [override](http://eslint.org/docs/rules/keyword-spacing#overrides) option we can use to disallow spaces after `catch`.

File: tests/fixtures/disallow-multiple-var-decl/good/for.js (hunk 1 above)
Comment (rwwagner90) [inline tests/fixtures/disallow-multiple-var-decl/good/for.js]: Yeah, I would expect all files to adhere to all other rules, except the one they are testing.

File: tests/fixtures/require-spread/good/super.js [hunk 4]
```diff
@@ -4,6 +4,7 @@ let Blah = Foo.extend({
 
     // odd, but still needed in some cases
     let { _super } = this;
+
```
Comment (brzpegasus) [inline tests/fixtures/require-spread/good/super.js]: I think we should remove the `newline-after-var` rule. I'm not sure how it made its way into the config in the first place, but I'd rather not introduce any new styling rule without discussing with the team first. And my personal opinion is that something like the following should be accepted since it's such a short block:

```js
function greeting(id) {
  let name = generateName(id);
  return `Hola, ${name}!`;
}
```

Comment (rwwagner90) [inline tests/fixtures/require-spread/good/super.js]: It was to enforce this https://github.com/DockYard/ember-suave/tree/master/tests/fixtures/rules/require-line-break-after-variable-assignment

Do we want to no longer enforce that?

Comment (rwwagner90) [inline tests/fixtures/require-spread/good/super.js]: I'm all for removing it. Just say the word!

File: tests/fixtures/disallow-multiple-var-decl/good/for.js (hunk 1 above)
Comment (brzpegasus) [inline tests/fixtures/disallow-multiple-var-decl/good/for.js]: I misread this. The `prefer-destructuring` rule should not complain about `j = arr.length` since we're assigning to a variable of a different name than the property. If it is, it's a bug with the rule.

File: tests/fixtures/require-spread/good/super.js (hunk 4 above)
Comment (rwwagner90) [inline tests/fixtures/require-spread/good/super.js]: I went ahead and removed it

Comment (brzpegasus) [inline tests/fixtures/require-spread/good/super.js]: I think that got taken care of by the `max-statements-per-line` rule, so I think we can nuke this.

File: tests/fixtures/require-template-strings-for-concatenation/good/example.js (hunk 2 above)
Comment (rwwagner90) [inline tests/fixtures/require-template-strings-for-concatenation/good/example.js]: I believe 'operator-linebreak': ['error', 'before'],

File: tests/fixtures/require-space-after-keywords/good/example.js (hunk 3 above)
Comment (rwwagner90) [inline tests/fixtures/require-space-after-keywords/good/example.js]: So you want spaces after all keywords, but not catch? Just trying to understand.

File: tests/fixtures/disallow-multiple-var-decl/good/for.js (hunk 1 above)
Comment (brzpegasus) [inline tests/fixtures/disallow-multiple-var-decl/good/for.js]: We can log ^ as a separate issue and address in a different PR if confirmed.

Comment (rwwagner90) [inline tests/fixtures/disallow-multiple-var-decl/good/for.js]: Yeah, I'm not sure why it doesn't like that, but it gives an error without those disable comments

File: tests/fixtures/require-space-after-keywords/good/example.js (hunk 3 above)
Comment (brzpegasus) [inline tests/fixtures/require-space-after-keywords/good/example.js]: Exactly.

**Comment by rwwagner90** [comment]
@brzpegasus how are we looking? Let me know anything else that needs to be addressed.

**Comment by brzpegasus** [comment]
I think the only thing left to adjust is the config to not allow spaces after `catch`.

**Comment by rwwagner90** [comment]
@brzpegasus just made that change. Let me know if anything else needs to be done

**Comment by brzpegasus** [comment]
:+1: Thank you!

**Comment by brzpegasus** [comment]
I can't merge PRs that have failing tests from a phone (didn't know that was a thing!). I'll get to it later from my laptop.

**Comment by rwwagner90** [comment]
Sounds good, thanks!

**Comment by rwwagner90** [comment]
@brzpegasus once you merge this in, can you also bump the version and release to npm? Would love to have this out and useable while we wrangle the last few issues.

**Comment by rwwagner90** [comment]
Thanks for the merge @brzpegasus!
